src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -0,0 +1,253 @@
+				return;
+
+			if (GetSymbolFromOperation (returnOperation) is not ISymbol returnedSymbol)
+				return;
+
+			returnedSymbol.TryGetDynamicallyAccessedMemberTypes (out var damtOnReturnedValue);
+			if (SourceHasMatchingAnnotations (damtOnReturnedValue, damtOnReturnType, out var missingAnnotations))
+				return;
+
+			var diag = GetDiagnosticId (returnedSymbol.Kind, context.ContainingSymbol.Kind, true);
+			var diagArgs = GetDiagnosticArguments (returnedSymbol.Kind == SymbolKind.NamedType ? context.ContainingSymbol : returnedSymbol, context.ContainingSymbol, missingAnnotations);
+			context.ReportDiagnostic (Diagnostic.Create (DiagnosticDescriptors.GetDiagnosticDescriptor (diag), returnOperation.Syntax.GetLocation (), diagArgs));
+		}
+
+		static void ProcessArguments (OperationAnalysisContext context, IInvocationOperation invocationOperation)
+		{
+			foreach (var argument in invocationOperation.Arguments) {
+				var targetParameter = argument.Parameter;
+				if (targetParameter is null || !targetParameter.TryGetDynamicallyAccessedMemberTypes (out var damtOnParameter))
+					return;
+
+				ISymbol sourceArgument = argument.Value.Kind == OperationKind.Conversion ? context.ContainingSymbol : GetSymbolFromOperation (argument.Value)!;
+				if (sourceArgument is null)
+					return;
+
+				sourceArgument.TryGetDynamicallyAccessedMemberTypes (out var damtOnArgument);
+				if (SourceHasMatchingAnnotations (damtOnArgument, damtOnParameter, out var missingAnnotations))
+					return;
+
+				var diag = GetDiagnosticId (sourceArgument.Kind, targetParameter.Kind, argument.Value.Kind == OperationKind.Conversion);
+				var diagArgs = GetDiagnosticArguments (sourceArgument, targetParameter, missingAnnotations);
+				context.ReportDiagnostic (Diagnostic.Create (DiagnosticDescriptors.GetDiagnosticDescriptor (diag), argument.Syntax.GetLocation (), diagArgs));
+			}
+		}
+
+		static void ProcessTypeArguments (OperationAnalysisContext context, IInvocationOperation invocationOperation)
+		{
+			var targetMethod = invocationOperation.TargetMethod;
+			for (int i = 0; i < targetMethod.TypeParameters.Length; i++) {
+				var arg = targetMethod.TypeArguments[i];
+				var param = targetMethod.TypeParameters[i];
+				if (!param.TryGetDynamicallyAccessedMemberTypes (out var damtOnTypeParameter))
+					continue;
+
+				arg.TryGetDynamicallyAccessedMemberTypes (out var damtOnTypeArgument);
+				if (SourceHasMatchingAnnotations (damtOnTypeArgument, damtOnTypeParameter, out var missingAnnotations))
+					continue;
+
+				var diag = GetDiagnosticId (arg.Kind, param.Kind);
+				var diagArgs = GetDiagnosticArguments (arg, param, missingAnnotations);
+				context.ReportDiagnostic (Diagnostic.Create (DiagnosticDescriptors.GetDiagnosticDescriptor (diag), invocationOperation.Syntax.GetLocation (), diagArgs));
+			}
+		}
+
+		static DiagnosticId GetDiagnosticId (SymbolKind source, SymbolKind target, bool targetsType = false)
```

Comment:
Yes, that's the intention. I think I'd better use another parameter name for this to make it clearer, and maybe add a comment. The logic for (Method, Parameter) was a bit off because of the `targetsType` was passed in ProcessArgument, I already changed that.